test/functional/ex_cmds/cd_spec.lua

```diff
@@ -0,0 +1,101 @@
+local gcwd = function() return eval('getcwd("g")') end  -- global dir
+local tcwd = function() return eval('getcwd("t")') end  -- tab dir
+local wcwd = function() return eval('getcwd("w")') end  -- window dir
+
+-- Test both the `cd` and `chdir` variants
+for _, cmd in ipairs {'cd', 'chdir'} do
+  describe(':*' .. cmd, function()
+    before_each(function()
+      clear()
+      for _, d in ipairs(directories) do
+        lfs.mkdir(d)
+      end
+    end)
+
+    after_each(function()
+      for _, d in ipairs(directories) do
+        lfs.rmdir(d)
+      end
+    end)
+
+    it('works', function()
+      -- Store the initial working directory
+      local globalDir = cwd()
+
+      -- Create a new tab first and verify that is has the same working dir
+      execute('tabnew')
+      eq(globalDir, cwd())
+      eq('', tcwd())  -- has neither tab- nor window-local directory
+      eq('', wcwd())
+
+      -- Change tab-local working directory and verify it is different
+      execute('t' .. cmd .. ' ' .. directories[1])
+      eq(globalDir .. '/' .. directories[1], cwd())
+      eq(cwd(), tcwd())  -- working directory maches tab directory
+      eq(   '', wcwd())  -- still no window-directory
+
+      -- Create a new window in this tab to test `:lcd`
+      execute('new')
+      eq(cwd(), tcwd())  -- Still tab-local working directory
+      eq(   '', wcwd())  -- Still no window-local working directory
+      eq(globalDir .. '/' .. directories[1], cwd())
+      execute('l' .. cmd .. ' ../' .. directories[2])
+      eq(globalDir .. '/' .. directories[2], cwd())
+      eq(globalDir .. '/' .. directories[1], tcwd())
+      eq(                             cwd(), wcwd())
+
+      -- Verify the first window still has the tab local directory
+      execute('wincmd w')
+      eq(globalDir .. '/' .. directories[1],  cwd())
+      eq(globalDir .. '/' .. directories[1], tcwd())
+      eq(                                '', wcwd())
+
+      -- Change back to initial tab and verify working directory has stayed
+      feed('gt')
+      eq(globalDir, cwd())
+      eq('', tcwd())
+      eq('', wcwd())
+
+      -- Verify global changes don't affect local ones
+      execute('' .. cmd .. ' ' .. directories[3])
+      eq(globalDir .. '/' .. directories[3], cwd())
+      feed('gt')
+      eq(globalDir .. '/' .. directories[1], cwd())
+      eq(globalDir .. '/' .. directories[1],  tcwd())
+      eq(                                '', wcwd())
+
+      -- Unless the global change happened in a tab with local directory
+      execute('' .. cmd .. ' ..')
+      eq(globalDir, cwd())
+      eq('', tcwd())
+      eq('', wcwd())
+      -- Which also affects the first tab
+      feed('gt')
+      eq(globalDir, cwd())
+
+      -- But not in a window with its own local directory
+      feed('gt')
+      execute('wincmd w')
+      eq(globalDir .. '/' .. directories[2], cwd())
+      eq(                                '', tcwd())
+      eq(globalDir .. '/' .. directories[2], wcwd())
```

Comment:
How do I test for errors and how do I test for invalid argument types? For longer strings only the first character is evaluated, so one could write `getcwd('tashdfkh')` and it would act like `getcwd('t')`, the only precondition is that the argument is either a string or nothing. Is that bad practice?